src/core/control/GeometryToolController.cpp

```diff
@@ -0,0 +1,118 @@
+#include "GeometryToolController.h"
+
+#include "control/Control.h"
+#include "control/layer/LayerController.h"
+#include "gui/XournalView.h"
+#include "gui/XournalppCursor.h"
+#include "model/GeometryTool.h"
+#include "model/Stroke.h"
+#include "model/XojPage.h"
+#include "undo/InsertUndoAction.h"
+
+using xoj::util::Rectangle;
+
+constexpr double MARK_SIZE = 2.;
+
+GeometryToolController::GeometryToolController(XojPageView* view, GeometryTool* geometryTool):
+        view(view), geometryTool(geometryTool) {}
+
+GeometryToolController::~GeometryToolController() = default;
+
+void GeometryToolController::move(double x, double y) {
+    geometryTool->setTranslationX(geometryTool->getTranslationX() + x);
+    geometryTool->setTranslationY(geometryTool->getTranslationY() + y);
+    geometryTool->notify();
+}
+
+void GeometryToolController::rotate(double da, double cx, double cy) {
+    geometryTool->setRotation(geometryTool->getRotation() + da);
+    const auto offsetX = geometryTool->getTranslationX() - cx;
+    const auto offsetY = geometryTool->getTranslationY() - cy;
+    const auto mx = offsetX * cos(da) - offsetY * sin(da);
+    const auto my = offsetX * sin(da) + offsetY * cos(da);
+    geometryTool->setTranslationX(cx + mx);
+    geometryTool->setTranslationY(cy + my);
+    geometryTool->notify();
+}
+
+void GeometryToolController::scale(double f, double cx, double cy) {
+    geometryTool->setHeight(geometryTool->getHeight() * f);
+    const auto offsetX = geometryTool->getTranslationX() - cx;
+    const auto offsetY = geometryTool->getTranslationY() - cy;
+    const auto mx = offsetX * f;
+    const auto my = offsetY * f;
+    geometryTool->setTranslationX(cx + mx);
+    geometryTool->setTranslationY(cy + my);
+    geometryTool->notify(true);
+}
+
+void GeometryToolController::markPoint(double x, double y) {
+    const auto control = view->getXournal()->getControl();
+    const auto h = control->getToolHandler();
+    Stroke* cross = new Stroke();
+    cross->setWidth(h->getToolThickness(TOOL_PEN)[TOOL_SIZE_FINE]);
+    cross->setColor(h->getTool(TOOL_PEN).getColor());
+    cross->addPoint(Point(x + MARK_SIZE, y + MARK_SIZE));
+    cross->addPoint(Point(x - MARK_SIZE, y - MARK_SIZE));
+    cross->addPoint(Point(x, y));
+    cross->addPoint(Point(x + MARK_SIZE, y - MARK_SIZE));
+    cross->addPoint(Point(x - MARK_SIZE, y + MARK_SIZE));
+
+    const auto page = view->getPage();
+    control->getLayerController()->ensureLayerExists(page);
+    const auto layer = page->getSelectedLayer();
+
+    const auto undo = control->getUndoRedoHandler();
+    undo->addUndoAction(std::make_unique<InsertUndoAction>(page, layer, cross));
+
+    layer->addElement(cross);
+
+    const Rectangle<double> rect{cross->getX(), cross->getY(), cross->getElementWidth(), cross->getElementHeight()};
+    view->rerenderRect(rect.x, rect.y, rect.width, rect.height);
+}
+
+void GeometryToolController::addStrokeToLayer() {
```

Comment:
Is the document's mutex locked when this is called? If not, then it needs to be locked from `ensureLayerExists` to `layer->addElement`

Reply:
Fine, I will lock it.

Reply:
Unfortunately the app freezes when I lock the document there.

Reply:
So it means the lock is either already locked by the caller (I could not find where), or one of the functions called tries to lock the alread locked mutex. We need to figure this out.

Reply:
You're right, the already locked mutex gets locked. I will investigate who locked it.

Reply:
The mutex when locked in the `GeometryToolController` gets locked again when calling
```
undo->addUndoAction(std::make_unique<InsertUndoAction>(page, layer, stroke.get()));
```
which calls `Control::updateWindowTitle()` via `UndoRedoHandler::fireUpdateUndoRedoButtons()` and `Control::undoRedoChanged()`.

If moving the `addUndoAction` past the mutex unlocking, there is no freeze anymore.

Reply:
Done in https://github.com/xournalpp/xournalpp/pull/4211/commits/1ff09f070b02d5f44cddd3a0bffcf62e15055a04

Reply:
Wow! This is a good example of bad code structure. The window title has nothing to do with the undo-redo stack. There should absolutely not be any call to `updateWindowTitle()` in `undoRedoChanged()`...

Reply:
When a document that has no changes compared to the saved version is modified, an asterix (*) is put into the window title (before the document name). When the modification is undone, the star is removed again. That's probably the reason, why the undo-redo stack was linked to the window title update.